# origin: Debian, #187962
# cset: n/a
# inclusion: not submitted
# description: cover up i810_dma problem, prevent oops
# revision date: 2004-09-03

diff -urN kernel-source-2.4.26/drivers/char/drm/i810_dma.c kernel-source-2.4.26-1/drivers/char/drm/i810_dma.c
--- kernel-source-2.4.26/drivers/char/drm/i810_dma.c	2004-02-19 00:36:31.000000000 +1100
+++ kernel-source-2.4.26-1/drivers/char/drm/i810_dma.c	2004-02-22 20:28:16.000000000 +1100
@@ -335,6 +335,10 @@
 
    	ring->head = I810_READ(LP_RING + RING_HEAD) & HEAD_ADDR;
      	ring->tail = I810_READ(LP_RING + RING_TAIL);
+     	if ((int)ring->tail < 0) {
+     		DRM_ERROR("buggy X server: ring->tail = %u\n", ring->tail);
+     		ring->tail = ring->head;
+     	}
      	ring->space = ring->head - (ring->tail+8);
      	if (ring->space < 0) ring->space += ring->Size;
 }
